### PR TITLE
Changed in memory repository to store json so it is more representati…

### PR DESCRIPTION
**Pull request by robertbennett1998**
Relevant issue - #145

### CODE REVIEW

File: Microsoft.Azure.CosmosRepository/test/Extensions/ItemsDictionaryEasyTryAdd.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) IEvangelist. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.CosmosRepositoryTests.Extensions
+{
+    public static class ItemsDictionaryEasyTryAdd
```
Comment (mumby0168) [inline Microsoft.Azure.CosmosRepository/test/Extensions/ItemsDictionaryEasyTryAdd.cs:10]: I would maybe just call this `ConcurrentDictionaryExtensions`

File: Microsoft.Azure.CosmosRepository/test/Extensions/ItemsDictionaryEasyTryAdd.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        public static bool TryAdd<TItem>(this ConcurrentDictionary<string, string> items, string id, TItem item)
+        {
```
Comment (mumby0168) [inline Microsoft.Azure.CosmosRepository/test/Extensions/ItemsDictionaryEasyTryAdd.cs:13]: This might be better named `TryAddAsJson` ?

**Review by mumby0168 — CHANGES_REQUESTED** [review]
Looks really good and certainly makes it closer to the real things, thanks 😃

**Review by mumby0168 — DISMISSED** [review]
Looks good to me we'll :shipit:, @IEvangelist are you happy with this?

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 3]
```diff
@@ -14,13 +14,18 @@
...
 
 namespace Microsoft.Azure.CosmosRepository
 {
     /// <inheritdoc/>
     internal class InMemoryRepository<TItem> : IRepository<TItem> where TItem : class, IItem
     {
-        internal ConcurrentDictionary<string, TItem> Items { get; } = new();
+        internal ConcurrentDictionary<string, string> Items { get; } = new();
+
+        internal string SerialiseItem(TItem item) => JsonConvert.SerializeObject(item);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:26]: For consistency with the delegated API that is being called, could we please rename all of these to `Serialize` and `Deserialize`? 😄 

```suggestion
        internal string SerializeItem(TItem item) => JsonConvert.SerializeObject(item);
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 4]
```diff
@@ -14,13 +14,18 @@
...
+        internal TItem DeserialiseItem(string serialisedItem) => JsonConvert.DeserializeObject<TItem>(serialisedItem);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:27]:
```suggestion
        internal TItem DeserializeItem(string jsonItem) => JsonConvert.DeserializeObject<TItem>(jsonItem);
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 5]
```diff
@@ -14,13 +14,18 @@
...
+        internal TDeserialiseTo DeserialiseItem<TDeserialiseTo>(string serialisedItem) => JsonConvert.DeserializeObject<TDeserialiseTo>(serialisedItem);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:28]:
```suggestion
        internal TDeserializeTo DeserializeItem<TDeserializeTo>(string jsonItem) => JsonConvert.DeserializeObject<TDeserializeTo>(jsonItem);
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 6]
```diff
@@ -36,7 +41,7 @@ public async ValueTask<TItem> GetAsync(string id, PartitionKey partitionKey, Can
                 partitionKey = new PartitionKey(id);
             }
 
-            TItem item = Items.Values.FirstOrDefault(i => i.Id == id && new PartitionKey(i.PartitionKey) == partitionKey);
+            TItem item = Items.Values.Select(DeserialiseItem).FirstOrDefault(i => i.Id == id && new PartitionKey(i.PartitionKey) == partitionKey);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:44]:
```suggestion
            TItem item = Items.Values.Select(DeserializeItem).FirstOrDefault(i => i.Id == id && new PartitionKey(i.PartitionKey) == partitionKey);
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 7]
```diff
@@ -50,7 +55,7 @@ public async ValueTask<TItem> GetAsync(string id, PartitionKey partitionKey, Can
         public async ValueTask<IEnumerable<TItem>> GetAsync(Expression<Func<TItem, bool>> predicate, CancellationToken cancellationToken = default)
         {
             await Task.CompletedTask;
-            return Items.Values.Where(predicate.Compose(
+            return Items.Values.Select(DeserialiseItem).Where(predicate.Compose(
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:58]:
```suggestion
            return Items.Values.Select(DeserializeItem).Where(predicate.Compose(
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 8]
```diff
@@ -73,16 +78,17 @@ public async ValueTask<TItem> CreateAsync(TItem value, CancellationToken cancell
 
             await Task.CompletedTask;
 
-            TItem item = Items.Values.FirstOrDefault(i => i.Id == value.Id && i.PartitionKey == value.PartitionKey);
+            TItem item = Items.Values.Select(DeserialiseItem).FirstOrDefault(i => i.Id == value.Id && i.PartitionKey == value.PartitionKey);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:81]:
```suggestion
            TItem item = Items.Values.Select(DeserializeItem).FirstOrDefault(i => i.Id == value.Id && i.PartitionKey == value.PartitionKey);
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 9]
```diff
@@ -73,16 +78,17 @@ public async ValueTask<TItem> CreateAsync(TItem value, CancellationToken cancell
...
             {
                 Conflict();
             }
 
-            Items.TryAdd(value.Id, value);
+            string serialisedValue = JsonConvert.SerializeObject(value);
+            Items.TryAdd(value.Id, serialisedValue);
 
-            return value;
+            return DeserialiseItem(Items[value.Id]);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:91]:
```suggestion
            return DeserializeItem(Items[value.Id]);
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 10]
```diff
@@ -103,9 +109,9 @@ public async ValueTask<TItem> UpdateAsync(TItem value, CancellationToken cancell
         {
             await Task.CompletedTask;
 
-            TItem item = Items.Values.FirstOrDefault(i => i.Id == value.Id && i.PartitionKey == value.PartitionKey);
+            Items[value.Id] = SerialiseItem(value);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:112]:
```suggestion
            Items[value.Id] = SerializeItem(value);
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 11]
```diff
@@ -103,9 +109,9 @@ public async ValueTask<TItem> UpdateAsync(TItem value, CancellationToken cancell
...
 
-            return Items[value.Id] = value;
+            return DeserialiseItem(Items[value.Id]);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:114]:
```suggestion
            return DeserializeItem(Items[value.Id]);
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 12]
```diff
@@ -131,7 +137,7 @@ public async ValueTask<IEnumerable<TItem>> UpdateAsync(IEnumerable<TItem> values
 
             partitionKeyValue ??= id;
 
-            IItem item = Items.Values.FirstOrDefault(x => x.Id == id && x.PartitionKey == partitionKeyValue);
+            TItem item = Items.Values.Select(DeserialiseItem).FirstOrDefault(x => x.Id == id && x.PartitionKey == partitionKeyValue);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:140]:
```suggestion
            TItem item = Items.Values.Select(DeserializeItem).FirstOrDefault(x => x.Id == id && x.PartitionKey == partitionKeyValue);
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 13]
```diff
@@ -169,7 +177,7 @@ public async ValueTask DeleteAsync(string id, PartitionKey partitionKey, Cancell
                 partitionKey = new PartitionKey(id);
             }
 
-            TItem item = Items.Values.FirstOrDefault(i => i.Id == id && new PartitionKey(i.PartitionKey) == partitionKey);
+            TItem item = Items.Values.Select(DeserialiseItem).FirstOrDefault(i => i.Id == id && new PartitionKey(i.PartitionKey) == partitionKey);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:180]:
```suggestion
            TItem item = Items.Values.Select(DeserializeItem).FirstOrDefault(i => i.Id == id && new PartitionKey(i.PartitionKey) == partitionKey);
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 14]
```diff
@@ -187,14 +195,14 @@ public ValueTask<bool> ExistsAsync(string id, string partitionKeyValue = null, C
         public async ValueTask<bool> ExistsAsync(string id, PartitionKey partitionKey, CancellationToken cancellationToken = default)
         {
             await Task.CompletedTask;
-            return Items.Values.FirstOrDefault(i => i.Id == id && new PartitionKey(i.PartitionKey) == partitionKey) is not null;
+            return Items.Values.Select(DeserialiseItem).FirstOrDefault(i => i.Id == id && new PartitionKey(i.PartitionKey) == partitionKey) is not null;
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:198]:
```suggestion
            return Items.Values.Select(DeserializeItem).FirstOrDefault(i => i.Id == id && new PartitionKey(i.PartitionKey) == partitionKey) is not null;
```

File: Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs [hunk 15]
```diff
@@ -187,14 +195,14 @@ public ValueTask<bool> ExistsAsync(string id, string partitionKeyValue = null, C
...
         }
 
         /// <inheritdoc/>
         public async ValueTask<bool> ExistsAsync(Expression<Func<TItem, bool>> predicate, CancellationToken cancellationToken = default)
         {
             await Task.CompletedTask;
-            return Items.Values.Any(predicate.Compose(
+            return Items.Values.Select(DeserialiseItem).Any(predicate.Compose(
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/InMemoryRepository.cs:205]:
```suggestion
            return Items.Values.Select(DeserializeItem).Any(predicate.Compose(
```

**Review by IEvangelist — APPROVED** [review]
This looks great, awesome job! I just had a few minor naming nits, otherwise - let's ship this.

**Comment by IEvangelist** [comment]
A few naming misses happened as a result of the change:

Error: InMemoryRepository.cs(158,25): error CS0103: The name 'SerialiseItem' does not exist in the current context [/home/runner/work/azure-cosmos-dotnet-repository/azure-cosmos-dotnet-repository/Microsoft.Azure.CosmosRepository/src/Microsoft.Azure.CosmosRepository.csproj]

Build FAILED.

Error: InMemoryRepository.cs(158,25): error CS0103: The name 'SerialiseItem' does not exist in the current context [/home/runner/work/azure-cosmos-dotnet-repository/azure-cosmos-dotnet-repository/Microsoft.Azure.CosmosRepository/src/Microsoft.Azure.CosmosRepository.csproj]

**Review by mumby0168 — APPROVED** [review]

**Pull request merged**